### PR TITLE
Mostleast

### PR DESCRIPTION
**Pull request by secondsun**
Added in karma most and karma least.

Also of note, karma is stored in a sorted set now.

### CODE REVIEW

File: bot.js [hunk 1]
```diff
@@ -17,7 +18,7 @@ if (process.env.AEROBOT_CONFIG) {
     var config = {
         irc:{
             host:'irc.freenode.net',
-            nick:'aerobot33',
+            nick:'summersbot',
```
Comment (kborchers) [inline bot.js]: Probably don't want to change the name here.

**Comment by kborchers** [comment]
Too many whitespace issues to list individually from my phone.

File: bot.js (hunk 1 above)
Comment (secondsun) [inline bot.js]: You saw nothing :)

File: bot.js [hunk 2]
```diff
@@ -60,9 +61,26 @@ ircConnection.addListener('message', function (from, to, message) {
     } else if(bot.isKarmaRequest(message)) {
         var key = "aerobot:karma:" + config.irc.host + ":" + util.normalizeChannelName(to);
         var request = bot.parseKarmaRequest(message);
-        client.hincrby(key, request.user, request.direction, function (err, reply) {
+        client.zincrby(key, request.direction, request.user, function (err, reply) {
             var actionText = request.direction === 1 ? 'gained' : 'lost';
             ircConnection.say(to, request.user + ' ' + actionText + ' a level! (Karma: ' + reply + ')');
         });
+    } else if (bot.isKarmaMostRequest(message)) {
+    var key = "aerobot:karma:" + config.irc.host + ":" + util.normalizeChannelName(to);
```
Comment (kborchers) [inline bot.js]: 69 - 75 tab in

File: bot.js [hunk 3]
```diff
@@ -60,9 +61,26 @@ ircConnection.addListener('message', function (from, to, message) {
...
+    client.zrevrange(key, 0, 5, 'WITHSCORES', function(err, reply) {
+        if (!err) {
+            var scores = util.normalizeKarmaScores(reply);
+                        ircConnection.say(to, util.prettyPrintKarmaScores(scores));
+        }    
+    });
+    } else if (bot.isKarmaLeastRequest(message)) {
+    var key = "aerobot:karma:" + config.irc.host + ":" + util.normalizeChannelName(to);
```
Comment (kborchers) [inline bot.js]: 77 - 83 tab in

File: lib/util.js [hunk 4]
```diff
@@ -120,3 +124,41 @@ function parseFactoids(payload, result) {
...
+            last = {};
+        }
+        if (last.user) {
+            last.score = parseInt(value)
+        } else {
+            last.user = value;
+        }
+        result.push(last);
+        return result;
+        },[]);
```
Comment (kborchers) [inline lib/util.js]: 1 tab out

File: test/util_spec.js [hunk 5]
```diff
@@ -13,6 +13,23 @@ describe('Util module', function() {
     it('knows how to normalize channel names', function () {
         util.normalizeChannelName('#aerobot-test').should.eql('aerobot-test');
     });
+    it('knows how to normalize karma sets', function () {
+        var scores = ['summersp','1','summersbot','0','summersbot1','-5'];
+        var normalScores = [{user:'summersp',score:1},
```
Comment (kborchers) [inline test/util_spec.js]: Yucky extra spaces and need more space in object literal. I would probably put those objects on multiple lines as well but at least need more spaces.

``` JS
var normalScores = [
    { user: 'summersp', score: 1 },
    { user: 'summersbot', score: 0 },
    { user: 'summersbot1', score: -5 }
];
```

File: test/util_spec.js [hunk 6]
```diff
@@ -13,6 +13,23 @@ describe('Util module', function() {
...
+                            {user:'summersbot',score:0},
+                            {user:'summersbot1',score:-5}];
+        util.normalizeKarmaScores(scores).should.eql(normalScores);
+    });
+    it('knows how to pretty print karma', function () {
+        var scores = [{user:'summersp',score:1},
```
Comment (kborchers) [inline test/util_spec.js]: Same as above

**Comment by qmx** [comment]
@secondsun it looks ok overally, could you rename the actions to karma best and karma worst?

**Comment by secondsun** [comment]
@qmx done

File: test/bot_spec.js [hunk 7]
```diff
@@ -44,6 +44,18 @@ describe('The Bot', function () {
         bot.isFactoidRetrievalRequest('?glwtd').should.be.ok;
         bot.isFactoidRetrievalRequest('? glwtd').should.not.be.ok;
     });
+    it('knows how to identify best karma request', function (){
+        var bot = new Bot('elephant');
+        bot.isKarmaBestRequest('karma best').should.be.ok;
+        bot.isKarmaWorstRequest('karma worst').should.be.ok;
+    });
+    it('knows how to identify factoid retrieval requests', function (){
```
Comment (qmx) [inline test/bot_spec.js]: copy paste #fail

**Comment by kborchers** [comment]
:+1: Except for one thing. We wanted best/worst, not most/least right? I think I saw you testing with that but don't see that change in the PR.

**Comment by secondsun** [comment]
@kborchers  see      4f5b675

**Comment by kborchers** [comment]
Weird, must have been cached. When I viewed the files, those changes weren't there.

**Pull request merged**